File: src/data/auditlog.ts
import { existsSync, readFileSync } from 'fs';
import { homedir } from 'os';
import { join } from 'path';
import type { OcAuditEntry } from '../types.js';

const AUDIT_LOG = join(homedir(), '.openclaw', 'logs', 'config-audit.jsonl');

function shortCommand(argv: string[]): string {
  // Find the openclaw subcommand from argv (skip node binary path)
  const idx = argv.findIndex(a => a.includes('openclaw') && !a.startsWith('--'));
  if (idx < 0) return argv.slice(1).join(' ');
  return argv.slice(idx).filter(a => !a.startsWith('--disable-warning')).join(' ');
}

export function loadAuditLog(): OcAuditEntry[] {
  if (!existsSync(AUDIT_LOG)) return [];

  const raw = readFileSync(AUDIT_LOG, 'utf-8').trim();
  if (!raw) return [];

  const entries: OcAuditEntry[] = [];
  for (const line of raw.split('\n')) {
    try {
      const obj = JSON.parse(line);
      const argv: string[] = obj.argv ?? [];
      const command = shortCommand(argv);
      entries.push({
        kind: 'auditlog',
        id: obj.ts,
        name: command,
        ts: obj.ts,
        event: obj.event ?? '',
        source: obj.source ?? '',
        configPath: obj.configPath ?? '',
        command,
        argv,
        pid: obj.pid ?? 0,
        result: obj.result ?? '',
        suspicious: obj.suspicious ?? [],
        previousBytes: obj.previousBytes ?? null,
        nextBytes: obj.nextBytes ?? 0,
        previousHash: obj.previousHash ?? '',
        nextHash: obj.nextHash ?? '',
        gatewayModeBefore: obj.gatewayModeBefore ?? null,
        gatewayModeAfter: obj.gatewayModeAfter ?? null,
      });
    } catch { /* skip malformed lines */ }
  }

  // Most recent first
  return entries.reverse();
}

File: internal/data/auditlog.go
package data

import (
	"bufio"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

func shortCommand(argv []string) string {
	for i, a := range argv {
		if strings.Contains(a, "openclaw") && !strings.HasPrefix(a, "--") {
			parts := argv[i:]
			var filtered []string
			for _, p := range parts {
				if !strings.HasPrefix(p, "--disable-warning") {
					filtered = append(filtered, p)
				}
			}
			return strings.Join(filtered, " ")
		}
	}
	if len(argv) > 1 {
		return strings.Join(argv[1:], " ")
	}
	return strings.Join(argv, " ")
}

func LoadAuditLog() []OcAuditEntry {
	home, _ := os.UserHomeDir()
	path := filepath.Join(home, ".openclaw", "logs", "config-audit.jsonl")

	f, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer f.Close()

	var entries []OcAuditEntry
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var obj struct {
			TS                string   `json:"ts"`
			Event             string   `json:"event"`
			Source            string   `json:"source"`
			ConfigPath        string   `json:"configPath"`
			Argv              []string `json:"argv"`
			PID               int      `json:"pid"`
			Result            string   `json:"result"`
			Suspicious        []string `json:"suspicious"`
			PreviousBytes     *int64   `json:"previousBytes"`
			NextBytes         int64    `json:"nextBytes"`
			PreviousHash      string   `json:"previousHash"`
			NextHash          string   `json:"nextHash"`
			GatewayModeBefore *string  `json:"gatewayModeBefore"`
			GatewayModeAfter  *string  `json:"gatewayModeAfter"`
		}
		if err := json.Unmarshal([]byte(line), &obj); err != nil {
			continue
		}
		cmd := shortCommand(obj.Argv)
		entries = append(entries, OcAuditEntry{
			ID:                obj.TS,
			Name_:             cmd,
			TS:                obj.TS,
			Event:             obj.Event,
			Source:            obj.Source,
			ConfigPath:        obj.ConfigPath,
			Command:           cmd,
			Argv:              obj.Argv,
			PID:               obj.PID,
			Result:            obj.Result,
			Suspicious:        obj.Suspicious,
			PreviousBytes:     obj.PreviousBytes,
			NextBytes:         obj.NextBytes,
			PreviousHash:      obj.PreviousHash,
			NextHash:          obj.NextHash,
			GatewayModeBefore: obj.GatewayModeBefore,
			GatewayModeAfter:  obj.GatewayModeAfter,
		})
	}

	// Reverse for most-recent-first
	for i, j := 0, len(entries)-1; i < j; i, j = i+1, j-1 {
		entries[i], entries[j] = entries[j], entries[i]
	}
	return entries
}
